[Fleet] Output Secrets Backend

hop-dev opened the pull request:
## Summary

Part of #157458

Adds the ability to sepcify secrets in outputs. Currently the following secrets are supported:

- kafka output SSL key
- kafka output password
- logstash output SSL key

The behaviour is as follows:

- on create, secrets are created and the plain string is replaced with a secret reference on the output saved object
- on update, if a secret is updated, the old secret is deleted and a new one is created, the new secret ref is added to the output
- on delete, all secrets are deleted

Secrets are only enabled if a fleet server of 8.10.0 or greater is connected.

Integration tests added for all scenarios.

hop-dev committed:
Create validation working

hop-dev committed:
secrets being created correctly

hop-dev committed:
password secret working

hop-dev committed:
output secrets added to agent policy

hop-dev committed:
update working

hop-dev committed:
delete secrets

hop-dev committed:
put create behind feature flah

hop-dev committed:
API docs

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `/oblt-deploy-serverless` : Deploy a serverless Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

hop-dev committed:
fix types

hop-dev committed:
add mock

hop-dev committed:
update SO snapshot

hop-dev committed:
fix deleted secrets test

hop-dev committed:
remove console.log

hop-dev committed:
do not throw if no settings object

hop-dev committed:
add mock logger

hop-dev committed:
renove assertion

hop-dev committed:
fix docs

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/check_mappings_update --fix'

hop-dev committed:
fix types

hop-dev committed:
remove unused res

hop-dev committed:
fix agent list resturn statement

hop-dev commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into output-secrets

hop-dev commented:
@elasticmachine merge upstream

hop-dev commented:
@elasticmachine merge upstream

hop-dev committed:
Merge branch 'main' into output-secrets

juliaElastic commented on x-pack/plugins/fleet/server/services/output.ts:
```diff
@@ -668,6 +685,11 @@ class OutputService {
       savedObjectType: OUTPUT_SAVED_OBJECT_TYPE,
     });
 
+    await deleteOutputSecrets({
```
I wonder if we should change the order, to delete the output first (line 693), and then the secrets? 
Now if the secret deletion succeeds, and the output deletion fails, the state will be inconsistent.

hop-dev replied:
Yes you're right 👍

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

juliaElastic commented on x-pack/plugins/fleet/server/services/secrets.ts:
```diff
@@ -237,6 +244,105 @@ export async function extractAndWriteSecrets(opts: {
+  };
+}
+
+function getOutputSecretPaths(
```
It would be nice to document in a readme, where we have to make changes to add a new output secret e.g. `remote-elasticsearch` service token

hop-dev commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into output-secrets

machadoum committed:
[Security Solutions] Fix entity analytics dashboard error that appear during cypress tests (#168709)

## Summary

Fix Entity analytics dashboard error that popped up during Cypress
tests.

Solutions:
I added the same check we have for the risk score fetch to the risk
score KPI fetch.
It prevents the UI from fetching the KPI when the index doesn't exist.

### Checklist

- [x] [Unit or functional
tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html)
were updated or added to match the most common scenarios

dhurley14 committed:
resets operator_users file to match production + risk score index. (#169102)

jedrazb committed:
[Enterprise Search][Connectors] Update extraction service UI restrictions of native connectors (#169104)

efegurkan committed:
[Enterprise Search]Render native connector docs conditionally (#169121)

## Summary

<img width="772" alt="Screenshot 2023-10-17 at 16 36 45"
src="https://github.com/elastic/kibana/assets/1410658/f668ab82-485d-46a6-8598-c1c9973b6255">

## Release Note
External documentation links are rendered conditionally now to avoid
empty links.

### Checklist
- [x] [Unit or functional
tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html)
were updated or added to match the most common scenarios

Commit:
[serverless] Give appex-qa team permissions to manage deployments (#169114)

nikitaindik committed:
[Security Solution] Test plan for prebuilt rule flyout (#167727)

**Resolves: https://github.com/elastic/kibana/issues/166160**

## Summary

Adds a test plan for the prebuilt rules flyout. Covers installation and
upgrade actions and also the content shown in Overview and Investigation
Guide tabs.

<img width="1301" alt="Scherm­afbeelding 2023-10-03 om 22 29 39"
src="https://github.com/elastic/kibana/assets/15949146/1aa7aa8b-3746-4767-82ad-f4910034d5eb">

---------

Co-authored-by: Georgii Gorbachev <georgii.gorbachev@elastic.co>

jbudz committed:
skip failing test suite (#169133)

stephmilovic committed:
[Security solution] Rename connector from AWS Bedrock to Amazon Bedrock (#169025)

angorayc committed:
[SecuritySolution][DataQualityDashboard] Migrate data quality dashboard APIs to versioned router (#169037)

## Summary

https://github.com/elastic/kibana/issues/168334
https://github.com/elastic/kibana/issues/166271


Before:

<img width="2560" alt="Screenshot 2023-10-16 at 21 45 02"
src="https://github.com/elastic/kibana/assets/6295984/2945179a-fb5b-4946-8b51-4769d20ad30f">
<img width="2558" alt="Screenshot 2023-10-16 at 21 45 09"
src="https://github.com/elastic/kibana/assets/6295984/254e00dd-bee9-4325-b418-e7ee481dea5d">

After:

<img width="2546" alt="Screenshot 2023-10-16 at 22 41 31"
src="https://github.com/elastic/kibana/assets/6295984/166a99a0-6f28-453a-a785-1197508170b5">

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>

qn895 committed:
[ML] Fix data drift calculating inaccurate p value when range is not of uniform distribution (#168757)

Commit:
Move Security's codegen step into the checks command as it has a short runtime (#168260)

## Summary

Security runs a code generator script on PRs. It only takes a few
seconds. This change moves it to the 'checks' command. This will save on
build agents.


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled
appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

Co-authored-by: Jon <jon@elastic.co>

nchaulet committed:
[Fleet] Fix integration sticky column after EUI upgrade (#169111)

yuliacech committed:
[Remote clusters] Fix empty states (#169015)

## Summary

I noticed that Remote clusters UI has a wrapper on empty states which is
probably a regression from the EuiPage deprecations migration. This PR
fixes the list and edit views. See screenshots below.

<details>

#### List empty state 

Before 
<img width="1919" alt="Screenshot 2023-10-16 at 19 27 12"
src="https://github.com/elastic/kibana/assets/6585477/e0e50958-c605-413c-b79b-80b6743ac6b0">

After
<img width="1918" alt="Screenshot 2023-10-16 at 19 42 24"
src="https://github.com/elastic/kibana/assets/6585477/7cc54494-fd0a-4e18-9ec9-cc501afbb978">

#### List error 

Before
<img width="1916" alt="Screenshot 2023-10-16 at 19 34 07"
src="https://github.com/elastic/kibana/assets/6585477/362169af-924e-40d0-9c25-d3d3fbabf112">

After
<img width="1918" alt="Screenshot 2023-10-16 at 19 45 18"
src="https://github.com/elastic/kibana/assets/6585477/c4931d64-e298-4962-9ba0-ddaf9a948a6a">


#### List error permissions 

Before 
<img width="1921" alt="Screenshot 2023-10-16 at 19 27 50"
src="https://github.com/elastic/kibana/assets/6585477/1eaf2fff-c8d6-45b7-b670-4b8dd4ab0844">

After 
<img width="1916" alt="Screenshot 2023-10-16 at 19 44 14"
src="https://github.com/elastic/kibana/assets/6585477/9bbcce09-9326-45a4-bb31-a42a13a999ac">

#### List loading 

Before
<img width="1918" alt="Screenshot 2023-10-16 at 20 08 54"
src="https://github.com/elastic/kibana/assets/6585477/11b2dab3-9e93-4e37-821c-fb367afa4fb9">

After
<img width="1918" alt="Screenshot 2023-10-16 at 20 10 18"
src="https://github.com/elastic/kibana/assets/6585477/30021cfe-cd8c-4aff-8877-f4b052ce3b7b">

#### Edit form error

Before
<img width="1918" alt="Screenshot 2023-10-16 at 19 59 54"
src="https://github.com/elastic/kibana/assets/6585477/1d54919f-f1fb-4938-93b3-f949705b12b2">

After
<img width="1918" alt="Screenshot 2023-10-16 at 20 03 59"
src="https://github.com/elastic/kibana/assets/6585477/a2369e1c-3e28-418e-a359-1564259c00cb">


#### Edit form loading 

Before
<img width="1918" alt="Screenshot 2023-10-16 at 20 00 34"
src="https://github.com/elastic/kibana/assets/6585477/ead45d19-43a3-4cc9-bd6a-338321b6757d">

After
<img width="1918" alt="Screenshot 2023-10-16 at 20 03 41"
src="https://github.com/elastic/kibana/assets/6585477/e563a54e-598c-42fc-88a9-fb1659a86fdd">


</details>

jbudz committed:
Fix storybooks (#169167)

https://buildkite.com/elastic/kibana-on-merge/builds/36947#018b3ee2-2f5a-4a69-9d88-8f074cd1b4fa/229-1431

Uses the fix found by mistic at
https://github.com/elastic/kibana/pull/168043

simianhacker committed:
[SLO] Add timeslice metric indicator (#168539)

## Summary

This PR adds the new Timeslice Metric indicator for SLOs. Due to the
nature of these statistical aggregations, this indicator requires the
budgeting method to be set to `timeslices`; we ignore the timeslice
threshold in favor of the threshold set in the metric definition.

Users can create SLOs based on the following aggregations:
- Average
- Min
- Max
- Sum
- Cardinality
- Percentile
- Document count
- Std. Deviation
- Last Value

Other notable feature include:

- The ability to define an equation which supports basic math and logic
- Users can define a threshold based on the equation for good slices vs
bad slices

<img width="800" alt="image"
src="https://github.com/elastic/kibana/assets/41702/05af1ced-40cd-4a05-9dfc-e83ea7bfb5ab">

### Counter Metric Example

<img width="800" alt="image"
src="https://github.com/elastic/kibana/assets/41702/05eb5a0f-3043-493d-8add-df0d3bf8de02">


CC: @lucasmoore

---------

Co-authored-by: Kevin Delemme <kdelemme@gmail.com>

mistic committed:
skip flaky suite (#168342)

mistic committed:
skip flaky suite (#168534)

mistic committed:
skip flaky suite (#168985)

mistic committed:
skip flaky suite (#169134)

mistic committed:
skip flaky suite (#169154)

mistic committed:
skip flaky suite (#167056)

jbudz committed:
[ci] Temporarily move osquery tests back to on_merge_unsupported pipeline (#169189)

These tests are currently failing - 
https://buildkite.com/elastic/kibana-on-merge/builds/36968
https://buildkite.com/elastic/kibana-on-merge/builds/36969

We can move these back to the main pipeline after tests are working and
[junit reports are
available](https://github.com/elastic/kibana/issues/169018).

kibanamachine committed:
[api-docs] 2023-10-18 Daily api_docs build (#169192)

Generated by
https://buildkite.com/elastic/kibana-api-docs-daily/builds/494

shahzad31 committed:
[Synthetics] Fix and refactor table row click actions for test runs table (#166915)

kertal committed:
[Discover] Unskip shared links test (#168923)

Improves and unskips Discover's shared links tests, and makes them a few seconds faster on top
Flaky test runner (100): https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/3515

vitaliidm committed:
[Security Solution][Detection Engine] updates ES|QL info popover (#168611)

## Summary

- addresses https://github.com/elastic/ux-writing/issues/41
- content of info tooltip is moved to doc page. Instead only link to
that page is displayed


### Before

<img width="1459" alt="Screenshot 2023-10-13 at 15 56 20"
src="https://github.com/elastic/kibana/assets/92328789/9a00ea01-19c1-4639-ad7b-9ae92829f075">


### After

<img width="1298" alt="Screenshot 2023-10-13 at 15 53 07"
src="https://github.com/elastic/kibana/assets/92328789/84811780-5d42-4ed3-b0d6-200467443e2f">


### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing
guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses
sentence case text and includes [i18n
support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x]
[Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html)
was added for features that require explanation or tutorials
- [x] [Unit or functional
tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html)
were updated or added to match the most common scenarios

---------

Co-authored-by: Nastasha Solomon <79124755+nastasha-solomon@users.noreply.github.com>

stratoula committed:
[ES|QL] Force ES client timeout after 2 minutes (#168929)

## Summary

We found out that for ES|QL queries while the kibana server (and
browser) timeouts in 2 minutes, the requests are still running in ES. So
instead of being aborted in 2 minutes, the hit a 5minute timeout from
the proxy which retries 3 times and then aborts.

This happens ONLY when bfetch is enabled (which is the default)

After an investigation with Rudolf it seems that we don't abort
correctly in bsearch. Lukas is fixing it here
https://github.com/elastic/kibana/pull/169041 .

shahzad31 committed:
[Synthetics] Service sync refactoring (#168587)

Co-authored-by: Justin Kambic <jk@elastic.co>

stratoula committed:
[ES|QL] Do not refresh the query when a user clicks the warning/error popover (#169193)

## Summary

Leftover from my ES|QL implementation. We don't need to rerun the query
when a user clicks the warning popover. I am removing this here.

Note: I still keep it on the footer component because I will need this
here https://github.com/elastic/kibana/pull/167754

Commit:
[Serverless] Add panels to side nav (#167774)

maxcold committed:
do not add fleet server to settings in serverless in start_fleet_server script (#169200)

## Summary

Starting fleet server fails when Kibana is running in serverless mode
with `Fleet server host write APIs are disabled`. Not adding the fleet
server to fleet settings fixes it. Kudos to @patrykkopycinski for the
fix!

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>

maryam-saeidi committed:
Add alerting_test_data package and a script to create rules for manual testing (#168493)

Closes https://github.com/elastic/actionable-observability/issues/140

## Summary

Add alerting_test_data package to create custom threshold and APM rules
for manual testing.

More information about the usage:
https://github.com/elastic/actionable-observability/pull/156

---------

Co-authored-by: almudenasanz <almudena.sanz@elastic.co>
Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>

hop-dev committed:
add feature flag

hop-dev committed:
delete secrets after saved object
